File: ex12/reconstruct.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/features2d.hpp>
#include <opencv2/calib3d.hpp>
#include <vector>
#include <string>
#include <fstream>
#include <filesystem>
#include <chrono>
#include <iomanip>
#include <sstream>

namespace fs = std::filesystem;

// Simple logging utility
class Logger {
public:
    enum Level { DEBUG, INFO, WARNING, ERROR };
    
    static void setLevel(Level level) { currentLevel = level; }
    
    template<typename... Args>
    static void debug(Args... args) { log(DEBUG, args...); }
    
    template<typename... Args>
    static void info(Args... args) { log(INFO, args...); }
    
    template<typename... Args>
    static void warning(Args... args) { log(WARNING, args...); }
    
    template<typename... Args>
    static void error(Args... args) { log(ERROR, args...); }

private:
    static Level currentLevel;
    
    static std::string getTimestamp() {
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        std::stringstream ss;
        ss << std::put_time(std::localtime(&time), "%Y-%m-%d %H:%M:%S");
        return ss.str();
    }
    
    static const char* getLevelString(Level level) {
        switch (level) {
            case DEBUG: return "DEBUG";
            case INFO: return "INFO";
            case WARNING: return "WARNING";
            case ERROR: return "ERROR";
            default: return "UNKNOWN";
        }
    }
    
    template<typename T, typename... Args>
    static void log(Level level, T first, Args... args) {
        if (level >= currentLevel) {
            std::cout << getTimestamp() << " - " << getLevelString(level) << " - " 
                     << first;
            ((std::cout << " " << args), ...);
            std::cout << std::endl;
        }
    }
};

Logger::Level Logger::currentLevel = Logger::INFO;

class PointCloudReconstructor {
private:
    std::vector<cv::Mat> images;
    cv::Ptr<cv::Feature2D> feature_detector;
    cv::Ptr<cv::DescriptorMatcher> matcher;
    
    struct CloudPoint {
        cv::Point3d pt;
        cv::Vec3b color;
    };
    
    std::vector<CloudPoint> point_cloud;

    bool loadImages(const std::string& path) {
        Logger::info("Loading images from:", path);
        
        if (fs::is_regular_file(path)) {
            Logger::info("Input is a video file");
            cv::VideoCapture cap(path);
            if (!cap.isOpened()) {
                Logger::error("Failed to open video file");
                return false;
            }
            
            cv::Mat frame;
            int frameCount = 0;
            while (cap.read(frame)) {
                images.push_back(frame.clone());
                frameCount++;
                if (frameCount % 10 == 0) {
                    Logger::info("Loaded", frameCount, "frames");
                }
            }
            Logger::info("Completed video loading: extracted", images.size(), "frames");
        } else {
            Logger::info("Input is a directory");
            int loadedCount = 0;
            for (const auto& entry : fs::directory_iterator(path)) {
                Logger::debug("Loading image:", entry.path().string());
                cv::Mat img = cv::imread(entry.path().string());
                if (!img.empty()) {
                    images.push_back(img);
                    loadedCount++;
                } else {
                    Logger::warning("Failed to load image:", entry.path().string());
                }
            }
            Logger::info("Loaded", loadedCount, "images from directory");
        }
        return !images.empty();
    }

    void detectAndMatch(const cv::Mat& img1, const cv::Mat& img2,
                       std::vector<cv::Point2f>& pts1, std::vector<cv::Point2f>& pts2) {
        Logger::debug("Detecting and matching features between image pair");
        
        std::vector<cv::KeyPoint> keypoints1, keypoints2;
        cv::Mat descriptors1, descriptors2;
        
        feature_detector->detectAndCompute(img1, cv::noArray(), keypoints1, descriptors1);
        feature_detector->detectAndCompute(img2, cv::noArray(), keypoints2, descriptors2);
        
        Logger::debug("Found", keypoints1.size(), "keypoints in first image and", 
                     keypoints2.size(), "in second image");
        
        std::vector<std::vector<cv::DMatch>> knn_matches;
        matcher->knnMatch(descriptors1, descriptors2, knn_matches, 2);
        
        // Ratio test
        std::vector<cv::DMatch> good_matches;
        for (const auto& match : knn_matches) {
            if (match[0].distance < 0.7 * match[1].distance) {
                good_matches.push_back(match[0]);
            }
        }
        
        pts1.clear();
        pts2.clear();
        for (const auto& match : good_matches) {
            pts1.push_back(keypoints1[match.queryIdx].pt);
            pts2.push_back(keypoints2[match.trainIdx].pt);
        }
        
        Logger::debug("Found", good_matches.size(), "good matches after ratio test");
    }

    void triangulatePoints(const cv::Mat& img1, const cv::Mat& img2,
                      const cv::Mat& R, const cv::Mat& t,
                      const std::vector<cv::Point2f>& pts1,
                      const std::vector<cv::Point2f>& pts2) {
    Logger::debug("Beginning point triangulation");
    size_t initial_size = point_cloud.size();
    
    // Camera matrix (assuming default parameters)
    double focal_length = std::max(img1.cols, img1.rows);
    cv::Point2d pp(img1.cols/2, img1.rows/2);
    cv::Mat K = (cv::Mat_<double>(3,3) << 
        focal_length, 0, pp.x,
        0, focal_length, pp.y,
        0, 0, 1);
    
    // Create a temporary matrix to store the result of hconcat
    cv::Mat Rt;
    cv::hconcat(R, t, Rt);  // Concatenate R and t horizontally
    
    // Projection matrices
    cv::Mat P1 = K * cv::Mat::eye(3, 4, CV_64F);
    cv::Mat P2 = K * Rt;  // Use the concatenated matrix
    
    cv::Mat points4D;
    cv::triangulatePoints(P1, P2, pts1, pts2, points4D);
    
    // Convert to 3D points
    int validPoints = 0;
    for (int i = 0; i < points4D.cols; i++) {
        cv::Mat x = points4D.col(i);
        x /= x.at<float>(3);
        CloudPoint cp;
        cp.pt = cv::Point3d(x.at<float>(0), x.at<float>(1), x.at<float>(2));
        
        // Get color from first image
        cv::Point2f pt = pts1[i];
        if (pt.x >= 0 && pt.x < img1.cols && pt.y >= 0 && pt.y < img1.rows) {
            cp.color = img1.at<cv::Vec3b>(pt);
            point_cloud.push_back(cp);
            validPoints++;
        }
    }
    
    Logger::debug("Added", validPoints, "new 3D points to cloud");
}

    void savePLY(const std::string& filename) {
        Logger::info("Saving point cloud to PLY file:", filename);
        Logger::info("Total points in cloud:", point_cloud.size());
        
        try {
            std::ofstream out(filename);
            if (!out.is_open()) {
                throw std::runtime_error("Unable to open file for writing");
            }
            
            out << "ply\n";
            out << "format ascii 1.0\n";
            out << "element vertex " << point_cloud.size() << "\n";
            out << "property float x\n";
            out << "property float y\n";
            out << "property float z\n";
            out << "property uchar red\n";
            out << "property uchar green\n";
            out << "property uchar blue\n";
            out << "end_header\n";
            
            for (const auto& point : point_cloud) {
                out << point.pt.x << " " << point.pt.y << " " << point.pt.z << " "
                    << (int)point.color[2] << " " << (int)point.color[1] << " " << (int)point.color[0] << "\n";
            }
            Logger::info("Successfully saved PLY file");
        } catch (const std::exception& e) {
            Logger::error("Failed to save PLY file:", e.what());
            throw;
        }
    }

public:
    PointCloudReconstructor() {
        feature_detector = cv::SIFT::create();
        matcher = cv::DescriptorMatcher::create(cv::DescriptorMatcher::FLANNBASED);
    }
    
    bool reconstruct(const std::string& input_path, const std::string& output_path) {
        Logger::info("Starting 3D reconstruction");
        try {
            if (!loadImages(input_path)) {
                Logger::error("Failed to load images");
                return false;
            }
            
            Logger::info("Processing", images.size(), "images");
            for (size_t i = 0; i < images.size() - 1; i++) {
                Logger::info("Processing image pair", i + 1, "/", images.size() - 1);
                
                std::vector<cv::Point2f> pts1, pts2;
                detectAndMatch(images[i], images[i+1], pts1, pts2);
                
                if (pts1.size() < 8) {
                    Logger::warning("Insufficient matching points (", pts1.size(), 
                                  ") for image pair", i + 1, ", skipping");
                    continue;
                }
                
                cv::Mat E = cv::findEssentialMat(pts1, pts2, 1.0, cv::Point2d(0,0));
                cv::Mat R, t;
                cv::recoverPose(E, pts1, pts2, R, t);
                
                triangulatePoints(images[i], images[i+1], R, t, pts1, pts2);
            }
            
            if (point_cloud.empty()) {
                Logger::error("Failed to reconstruct point cloud - no points generated");
                return false;
            }
            
            savePLY(output_path);
            Logger::info("Reconstruction completed successfully");
            return true;
            
        } catch (const std::exception& e) {
            Logger::error("Reconstruction failed with error:", e.what());
            throw;
        }
    }
};

int main(int argc, char** argv) {
    std::string input_path, output_path;
    bool debug_mode = false;
    
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "-i" && i + 1 < argc) {
            input_path = argv[++i];
        } else if (arg == "-o" && i + 1 < argc) {
            output_path = argv[++i];
        } else if (arg == "--debug") {
            debug_mode = true;
        }
    }
    
    if (input_path.empty() || output_path.empty()) {
        std::cerr << "Usage: " << argv[0] << " -i <input_path> -o <output_path> [--debug]\n";
        return 1;
    }
    
    if (debug_mode) {
        Logger::setLevel(Logger::DEBUG);
    }
    
    try {
        PointCloudReconstructor reconstructor;
        if (!reconstructor.reconstruct(input_path, output_path)) {
            return 1;
        }
    } catch (const std::exception& e) {
        Logger::error("Fatal error:", e.what());
        return 1;
    }
    
    return 0;
}

File: ex12/reconstruct.py
import cv2
import numpy as np
import argparse
from pathlib import Path
import os
import logging

class PointCloudReconstructor:
    def __init__(self):
        self.images = []
        self.feature_detector = cv2.SIFT_create()
        self.matcher = cv2.FlannBasedMatcher_create()
        self.point_cloud = []  # List of (point, color) tuples
        
        # Setup logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )
        self.logger = logging.getLogger(__name__)
        
    def load_images(self, path):
        self.logger.info(f"Loading images from: {path}")
        path = Path(path)
        if path.is_file():  # Video file
            self.logger.info("Input is a video file")
            cap = cv2.VideoCapture(str(path))
            if not cap.isOpened():
                self.logger.error("Failed to open video file")
                return False
                
            frame_count = 0
            while True:
                ret, frame = cap.read()
                if not ret:
                    break
                self.images.append(frame)
                frame_count += 1
                if frame_count % 10 == 0:  # Log every 10th frame
                    self.logger.info(f"Loaded {frame_count} frames")
            cap.release()
            self.logger.info(f"Completed video loading: extracted {len(self.images)} frames")
        else:  # Directory of images
            self.logger.info("Input is a directory")
            valid_extensions = {'.jpg', '.jpeg', '.png', '.bmp'}
            for file in path.iterdir():
                if file.suffix.lower() in valid_extensions:
                    self.logger.debug(f"Loading image: {file.name}")
                    img = cv2.imread(str(file))
                    if img is not None:
                        self.images.append(img)
                    else:
                        self.logger.warning(f"Failed to load image: {file.name}")
            self.logger.info(f"Loaded {len(self.images)} images from directory")
                        
        return len(self.images) > 0
        
    def detect_and_match(self, img1, img2):
        self.logger.debug("Detecting and matching features between image pair")
        # Detect keypoints and compute descriptors
        kp1, des1 = self.feature_detector.detectAndCompute(img1, None)
        kp2, des2 = self.feature_detector.detectAndCompute(img2, None)
        
        self.logger.debug(f"Found {len(kp1)} keypoints in first image and {len(kp2)} in second image")
        
        # Match features
        matches = self.matcher.knnMatch(des1, des2, k=2)
        
        # Apply ratio test
        good_matches = []
        pts1 = []
        pts2 = []
        
        for m, n in matches:
            if m.distance < 0.7 * n.distance:
                good_matches.append(m)
                pts1.append(kp1[m.queryIdx].pt)
                pts2.append(kp2[m.trainIdx].pt)
        
        self.logger.debug(f"Found {len(good_matches)} good matches after ratio test")
                
        return np.float32(pts1), np.float32(pts2)
        
    def triangulate_points(self, img1, img2, R, t, pts1, pts2):
        self.logger.debug("Beginning point triangulation")
        # Camera matrix (assuming default parameters)
        focal_length = max(img1.shape[1], img1.shape[0])
        pp = (img1.shape[1]//2, img1.shape[0]//2)
        K = np.array([
            [focal_length, 0, pp[0]],
            [0, focal_length, pp[1]],
            [0, 0, 1]
        ])
        
        # Projection matrices
        P1 = K @ np.hstack((np.eye(3), np.zeros((3,1))))
        P2 = K @ np.hstack((R, t))
        
        # Triangulate
        points4D = cv2.triangulatePoints(P1, P2, pts1.T, pts2.T)
        points4D /= points4D[3]
        points3D = points4D[:3].T
        
        initial_point_count = len(self.point_cloud)
        # Get colors from first image
        for i, (pt3d, pt2d) in enumerate(zip(points3D, pts1)):
            x, y = int(pt2d[0]), int(pt2d[1])
            if 0 <= x < img1.shape[1] and 0 <= y < img1.shape[0]:
                color = img1[y, x]
                self.point_cloud.append((pt3d, color))
        
        new_points = len(self.point_cloud) - initial_point_count
        self.logger.debug(f"Added {new_points} new 3D points to cloud")
                
    def save_ply(self, filename):
        self.logger.info(f"Saving point cloud to PLY file: {filename}")
        self.logger.info(f"Total points in cloud: {len(self.point_cloud)}")
        
        try:
            with open(filename, 'w') as f:
                f.write("ply\n")
                f.write("format ascii 1.0\n")
                f.write(f"element vertex {len(self.point_cloud)}\n")
                f.write("property float x\n")
                f.write("property float y\n")
                f.write("property float z\n")
                f.write("property uchar red\n")
                f.write("property uchar green\n")
                f.write("property uchar blue\n")
                f.write("end_header\n")
                
                for point, color in self.point_cloud:
                    f.write(f"{point[0]} {point[1]} {point[2]} "
                           f"{int(color[2])} {int(color[1])} {int(color[0])}\n")
            self.logger.info("Successfully saved PLY file")
        except Exception as e:
            self.logger.error(f"Failed to save PLY file: {str(e)}")
            raise
                
    def reconstruct(self, input_path, output_path):
        self.logger.info("Starting 3D reconstruction")
        if not self.load_images(input_path):
            self.logger.error("Failed to load images")
            return False
            
        self.logger.info(f"Processing {len(self.images)} images")
        for i in range(len(self.images) - 1):
            self.logger.info(f"Processing image pair {i+1}/{len(self.images)-1}")
            pts1, pts2 = self.detect_and_match(self.images[i], self.images[i+1])
            
            if len(pts1) < 8:
                self.logger.warning(f"Insufficient matching points ({len(pts1)}) for image pair {i+1}, skipping")
                continue
                
            # Find essential matrix and recover pose
            E, mask = cv2.findEssentialMat(pts1, pts2, focal=1.0, pp=(0,0))
            _, R, t, mask = cv2.recoverPose(E, pts1, pts2)
            
            # Triangulate points
            self.triangulate_points(self.images[i], self.images[i+1], R, t, pts1, pts2)
            
        if not self.point_cloud:
            self.logger.error("Failed to reconstruct point cloud - no points generated")
            return False
            
        self.save_ply(output_path)
        self.logger.info("Reconstruction completed successfully")
        return True

def main():
    parser = argparse.ArgumentParser(description='3D reconstruction from images/video')
    parser.add_argument('-i', required=True, help='Input video file or image directory')
    parser.add_argument('-o', required=True, help='Output PLY file')
    parser.add_argument('--debug', action='store_true', help='Enable debug logging')
    args = parser.parse_args()
    
    if args.debug:
        logging.getLogger().setLevel(logging.DEBUG)
    
    reconstructor = PointCloudReconstructor()
    try:
        if reconstructor.reconstruct(args.i, args.o):
            logging.info("Reconstruction completed successfully")
        else:
            logging.error("Reconstruction failed")
    except Exception as e:
        logging.error(f"Reconstruction failed with error: {str(e)}")
        raise

if __name__ == '__main__':
    main()
